tools/AutoTuner/src/autotuner/utils.py

```diff
@@ -0,0 +1,652 @@
+        if stage_name == "finish" and "power__total" in value:
+            total_power = value["power__total"]
+        if stage_name == "finish" and "design__instance__utilization" in value:
+            final_util = value["design__instance__utilization"]
+        if stage_name == "finish" and "design__instance__area" in value:
+            design_area = value["design__instance__area"]
+        if stage_name == "finish" and "design__core__area" in value:
+            core_area = value["design__core__area"]
+        if stage_name == "finish" and "design__die__area" in value:
+            die_area = value["design__die__area"]
+    ret = {
+        "clk_period": clk_period,
+        "worst_slack": worst_slack,
+        "total_power": total_power,
+        "core_util": core_util,
+        "final_util": final_util,
+        "design_area": design_area,
+        "core_area": core_area,
+        "die_area": die_area,
+        "wirelength": wirelength,
+        "num_drc": num_drc,
+    }
+    return ret
+
+
+def read_config(file_name, mode, algorithm):
+    """
+    Please consider inclusive, exclusive
+    Most type uses [min, max)
+    But, Quantization makes the upper bound inclusive.
+    e.g., qrandint and qlograndint uses [min, max]
+    step value is used for quantized type (e.g., quniform). Otherwise, write 0.
+    When min==max, it means the constant value
+    """
+
+    def read(path):
+        # if file path does not exist, return empty string
+        print(os.path.abspath(path))
+        if not os.path.isfile(os.path.abspath(path)):
+            return ""
+        with open(os.path.abspath(path), "r") as file:
+            ret = file.read()
+        return ret
+
+    def read_sweep(this):
+        return [*this["minmax"], this["step"]]
+
+    def apply_condition(config, data):
+        from ray import tune
+        import random
+
+        # TODO: tune.sample_from only supports random search algorithm.
+        # To make conditional parameter for the other algorithms, different
+        # algorithms should take different methods (will be added)
+        if algorithm != "random":
+            return config
+        dp_pad_min = data["CELL_PAD_IN_SITES_DETAIL_PLACEMENT"]["minmax"][0]
+        dp_pad_step = data["CELL_PAD_IN_SITES_DETAIL_PLACEMENT"]["step"]
+        if dp_pad_step == 1:
+            config["CELL_PAD_IN_SITES_DETAIL_PLACEMENT"] = tune.sample_from(
+                lambda spec: np.random.randint(
+                    dp_pad_min, spec.config.CELL_PAD_IN_SITES_GLOBAL_PLACEMENT + 1
+                )
+            )
+        if dp_pad_step > 1:
+            config["CELL_PAD_IN_SITES_DETAIL_PLACEMENT"] = tune.sample_from(
+                lambda spec: random.randrange(
+                    dp_pad_min,
+                    spec.config.CELL_PAD_IN_SITES_GLOBAL_PLACEMENT + 1,
+                    dp_pad_step,
+                )
+            )
+        return config
+
+    def read_tune(this):
+        from ray import tune
```

Comment:
`ray.tune()` is pretty much used in most of the parsing anyways. is it a big overhead?

Reply:
It's already prepared for #2428, so importing `ray.tune` is not required when using functions from utils.

As Python does not import modules more than once, this line introduces minimal overhead at most.